File: python/RotateList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def rotateRight(self, head, k):
        if head is None or head.next is None:
            return head
        
        if k == 0:
            return head

        fast = head
        slow = head
        for i in xrange(k):
            if fast is None:
                fast = head
            fast = fast.next
                
        if fast is None:
            return head
        
        while fast.next is not None:
            fast = fast.next
            slow = slow.next

        result = slow.next
        slow.next = None
        fast.next = head
        
        return result

if __name__ == "__main__":
    solution = Solution()
    l1 = ListNode(1)
    l2 = ListNode(2)
    l3 = ListNode(3)
    l4 = ListNode(4)
    l5 = ListNode(5)
    l1.next = l2
    #l2.next = l3
    #l3.next = l4
    #l4.next = l5

    p = solution.rotateRight(l1, 3)
    while p is not None:
        print p.val
        p = p.next

File: java/RotateList.java
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        int length = getLength(head);
        if (length == 0) {
            return null;
        }
        
        int realK = k % length;
        if (realK == 0) {
            return head;
        }
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        int i = 0;
        while (i < realK) {
            i++;
            fast = fast.next;
        }
        
        ListNode slow = dummy;
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        
        ListNode temp = slow.next;
        slow.next = null;
        fast.next = dummy.next;
        dummy.next = temp;
        return dummy.next;
    }
    
    private int getLength(ListNode head) {
        ListNode p = head;
        int result = 0;
        
        while (p != null) {
            p = p.next;
            result ++;
        }
        
        return result;
    }
 }
